### PR TITLE
javascript

### PR DESCRIPTION
**Pull request by tpetr**
This PR contains:
1. a new dune project (`js`) that uses js_of_ocaml to create a javascript module for running semgrep
2. js stubs for pcre, tree-sitter, and our languages
3. minor tweaks to make semgrep run in the browser

TODO:
- Update this PR (or open additional PR) to include the machinery for building the companion `.wasm` file
- Add instructions on how to test this out locally

Any and all feedback welcome, but my main ask right now is a sanity check that my ocaml code & dune configuration is idiomatic, and that I didn't miss simpler/cleaner ways to do things.

PR checklist:

- [ ] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [ ] Tests included or PR comment includes a reproducible test plan
- [ ] Documentation is up-to-date
- [ ] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [ ] Change has no security implications (otherwise, ping security team)

If you're unsure on any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

File: libs/commons/Common.ml [hunk 1]
```diff
@@ -828,21 +828,27 @@ let cat file =
...
-        if Buffer.length extbuf >= max_len then Buffer.sub extbuf 0 max_len
-        else loop fd
-  in
-  let fd = Unix.openfile path [ Unix.O_RDONLY ] 0 in
-  Fun.protect ~finally:(fun () -> Unix.close fd) (fun () -> loop fd)
+  if !jsoo then (
+    let ic = open_in path in
+    let s = really_input_string ic (in_channel_length ic) in
+    close_in ic;
+    s)
```
Comment (tpetr) [inline libs/commons/Common.ml]: We're going to need to make a similar change to https://github.com/returntocorp/ocaml-tree-sitter-core/blob/main/src/run/lib/Util_file.ml#L25-L41

**Review by aryx — APPROVED** [review]

File: src/js/semgrep.js [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+//Provides: octs_create_parser_c_sharp
+//Requires: create_parser, tree_sitter_wasm_module
+function octs_create_parser_c_sharp() {
+    return create_parser(tree_sitter_wasm_module._tree_sitter_c_sharp())
+}
+
+//Provides: octs_create_parser_julia
+//Requires: create_parser
+function octs_create_parser_julia() {
+    throw new Error("TODO: re-incorporate julia once we figure out why the build is so slow") // TODO
```
Comment (tpetr) [inline src/js/semgrep.js]: @brandonspark 👋

**Comment by aryx** [comment]
for pre-commit you can look at https://semgrep.dev/docs/contributing/semgrep-core-contributing/
mostly need
pre-commit install
pre-commit run -a
to check everything works.
You will probably need to install ocamlformat too. opam install --deps-only dev/dev.opam should work.

**Comment by aryx** [comment]
```
make[1]: Entering directory '/__w/semgrep/semgrep'
dune build
File "src/js/dune", line 3, characters 58-73:
3 |  (libraries semgrep.running js_of_ocaml integers_stubs_js ctypes_stubs_js)
                                                              ^^^^^^^^^^^^^^^
Error: Library "ctypes_stubs_js" not found.
-> required by _build/default/src/js/Main.bc-for-jsoo
-> required by alias src/js/all
-> required by alias default
File "src/js/dune", line 14, characters 7-22:
14 |   (pps js_of_ocaml-ppx js_of_ocaml-ppx_deriving_json)))
            ^^^^^^^^^^^^^^^
Error: Library "js_of_ocaml-ppx" not found.
```

**Comment by aryx** [comment]
probably need to update our dependencies to support those packages.
you can probably add a few packages in semgrep.opam at the root; this will be used by 'make setup' in CI
to install all the necessary dependencies.

**Comment by tpetr** [comment]
Yep you're right, I must have muddied my opam environment during testing. Will fix.

**Comment by aryx** [comment]
can you modify pcre.js so that it parses with their current JS parser? (fun that they actually use the js parser I wrote in semgrep :) )

**Comment by tpetr** [comment]
> can you modify pcre.js so that it parses with their current JS parser?

Done. Turns out there were a couple places where I forgot to convert between JS and caml strings (and the latest jsoo is apparently more lenient about this) -- should be good now.

**Review by aryx — APPROVED** [review]

**Comment by aryx** [comment]
next steps: add stuff to test this in CI. Add tests so can check whether things work in the terminal with nodejs,
add instructions to get the wasm files.

**Pull request merged**